openedx/adg/lms/applications/test/test_views.py

```diff
@@ -282,3 +283,222 @@ def test_get_no_user_application_exists_for_application_success_view(get_request
+        cover_letter_view_get_request, 'adg/lms/applications/cover_letter.html', expected_context
+    )
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('btn_next, template', [('back', 'application_experience'), ('submit', 'application_hub')],
+                         ids=['back', 'submit'])
+def test_post_back_or_submit_written_application_cover_letter_view(btn_next, template, cover_letter_view_post_request):
+    # pylint: disable=protected-access
+    """
+    Test the case when the user either clicks on back or submit button
+    """
+    _mutable = cover_letter_view_post_request.POST._mutable
+    cover_letter_view_post_request.POST._mutable = True
+    cover_letter_view_post_request.POST['next'] = btn_next
+    cover_letter_view_post_request.POST._mutable = _mutable
+
+    response = CoverLetterView.as_view()(cover_letter_view_post_request)
+
+    assert response.get('Location') == reverse(template)
+    assert response.status_code == 302
+
+
+@pytest.mark.django_db
+def test_post_with_business_line_cover_letter_view(cover_letter_view_post_request):
+    # pylint: disable=protected-access
+    """
+    Test the case when the user has selected business line
+    """
+    business_line = BusinessLineFactory()
+
+    _mutable = cover_letter_view_post_request.POST._mutable
+    cover_letter_view_post_request.POST._mutable = True
+    cover_letter_view_post_request.POST['business_line'] = business_line.id
+    cover_letter_view_post_request.POST['next'] = 'back'
+    cover_letter_view_post_request.POST._mutable = _mutable
+
+    response = CoverLetterView.as_view()(cover_letter_view_post_request)
+
+    assert response.get('Location') == reverse('application_experience')
+    assert response.status_code == 302
+
+
+@pytest.mark.django_db
+@mock.patch('openedx.adg.lms.applications.views.UserApplication.save')
+@pytest.mark.parametrize('cover_letter', ['cover letter', None], ids=['typed_cover_letter', 'no_cover_letter'])
+def test_post_with_cover_letter_typed_cover_letter_view(mock_save, cover_letter, cover_letter_view_post_request):
+    # pylint: disable=protected-access
+    """
+    Test the case when the user has typed a cover letter or neither typed a cover letter nor uploaded a file
+    """
+    mock_save.return_value = None
+
+    _mutable = cover_letter_view_post_request.POST._mutable
+    cover_letter_view_post_request.POST._mutable = True
+    cover_letter_view_post_request.POST['text-coverletter'] = cover_letter
+    cover_letter_view_post_request.POST['next'] = 'back'
+    cover_letter_view_post_request.POST._mutable = _mutable
+
+    response = CoverLetterView.as_view()(cover_letter_view_post_request)
+
+    assert response.get('Location') == reverse('application_experience')
+    assert response.status_code == 302
+
+
+@pytest.mark.django_db
+@mock.patch('openedx.adg.lms.applications.views.UserApplication.save')
+def test_post_with_cover_letter_file_upload_cover_letter_view(mock_save, cover_letter_view_post_request):
```

Comment:
We can merge this function with the second last function using parameterization to avoid code repetition, and the logic is also similar. All you need to to do is add another parameter 'cover letter file', and based on that set request.FILES. The rest of the logic stays the same. Parameters are already used quite nicely in the 2nd last function, that too with mocking. Good effort.